### ISSUE TITLE
Ty misses required-field diagnostic for SQLModel (works for BaseModel)

### ISSUE THREAD

**Issue by Duckling92** [astral-sh/ty#1329]
### Summary
Ty flags missing required fields for `pydantic.BaseModel` but not for `sqlmodel.SQLModel` subclasses, both using Pydantic v2.

### Minimal repro
```py
from __future__ import annotations

from pydantic import BaseModel
from sqlmodel import SQLModel


class ExamplePydantic(BaseModel):
    number: float


class ExampleSQLModel(SQLModel):
    number: float


ExampleSQLModel()  # ty: no error (bug)
ExamplePydantic()  # ty: error (good)
```

### Expected
Expected: Both `ExampleSQLModel()` and `ExamplePydantic()` are flagged as missing required field number.
Actual: Only `ExamplePydantic()` is flagged, `ExampleSQLModel()` passes.

### Environment
python: 3.11.10
sqlmodel: 0.0.27
OS: MacOS Tahoe (26.0.1) 

### Version

ty 0.0.1-alpha.21 (ef52a1940 2025-09-19)

**Comment by carljm** [comment]
I am pretty sure the problem here is that `SQLModel` defines `__new__` and `__init__` with very permissive signatures: https://github.com/fastapi/sqlmodel/blob/main/sqlmodel/main.py#L789-L799

And I think if we see an explicitly-defined `__new__` or `__init__`, we assume that takes precedence over the dataclass-transform-generated one. So we validate the constructor call against the permissive `__new__` and `__init__` signatures and see no error.

But in this case it looks like SQLModel is just passing the arguments through, so its `__new__` is just a wrapper, and the preferred behavior would be to still validate the arguments against the dataclass-transform-generated constructor.

But it's a little tricky to determine under exactly which conditions we should make this assumption (that a constructor with permissive signature is just a pass-through.)

**Comment by carljm** [comment]
For reference, mypy and pyrefly both have the same behavior we do here (catching only the `ExamplePydantic()` error). Pyright catches both errors.

@erictraut do you happen to remember what heuristic or algorithm pyright uses here to decide that a custom constructor should be effectively ignored?

**Comment by erictraut** [comment]
Pyright isn't ignoring the custom `__new__`. It's overriding it with a synthesized `__new__`. Pyright synthesizes both an `__init__` and a `__new__` method for dataclasses. In the case of `__new__`, it synthesizes a method with the signature `def __new__(cls: type[<class>], *args: Any, *kwargs: Any) -> <class>`. If it didn't synthesize a `__new__` and the dataclass didn't provide a custom `__new__`, then all calls to the constructor would be flagged as an error because `object.__new__` would be used, and it doesn't support any parameters.

I'm not convicted that pyright's behavior is correct here.

It looks like ty is not synthesizing a `__new__`. If that's the case, how do you avoid the problem with `object.__new__`? I'm guessing that you're using a heuristic to skip this check?

```python
from dataclasses import dataclass

@dataclass
class DC:
    number: int

reveal_type(DC.__new__) # def __new__(cls) -> Self@new
DC.__new__(DC, 1) # ty reports an error here
DC(1) # ty does not report an error here
```

Regardless, it would be preferable for `SQLModel` to have less permissive type annotations here.

**Comment by sharkdp** [comment]
> It looks like ty is not synthesizing a `__new__`. If that's the case, how do you avoid the problem with `object.__new__`? I'm guessing that you're using a heuristic to skip this check?

Correct, we skip `object.__new__` when `__new__` is called implicitly: https://github.com/astral-sh/ruff/blob/75f3c0e8e6dcd52e8194f9831aef25fddb646685/crates/ty_python_semantic/src/types.rs#L3252

**Comment by carljm** [comment]
> then all calls to the constructor would be flagged as an error because `object.__new__` would be used, and it doesn't support any parameters.

But this problem with the signature of `object.__new__` is not specific to dataclasses, so "synthesizing a `__new__` method for dataclasses" wouldn't be a general solution to the problem. Does pyright actually synthesize a `__new__` method in that form for all classes that would otherwise inherit `object.__new__`?

IIRC ty's approach (skipping `object.__new__`) is roughly based on what happens at runtime, which is that `object.__new__` actually [only enforces its signature when called on a type that overrides `__new__`](https://github.com/python/cpython/blob/main/Objects/typeobject.c#L7108) or that doesn't override `__init__`, otherwise it accepts any and all arguments without complaint.

**Comment by theelderbeever** [comment]
Possibly related but `pydantic` is reporting missing required params as well which shouldn't need defined
> No arguments provided for required parameters `__pydantic_extra__`, `__pydantic_fields_set__`, `__pydantic_private__` (ty missing-argument)

**Comment by sharkdp** [comment]
> Possibly related but `pydantic` is reporting missing required params as well which shouldn't need defined
> 
> > No arguments provided for required parameters `__pydantic_extra__`, `__pydantic_fields_set__`, `__pydantic_private__` (ty missing-argument)

I don't think that's related. Let's discuss pydantic separately (in https://github.com/astral-sh/ty/issues/1421 or in a new ticket, if you think this is a distinct problem from #1431)

**Comment by carljm** [comment]
Given that our behavior here matches both mypy and pyrefly, and is correct given the annotations on `SQLModel.__new__`, I'm inclined to say there's nothing to be done here. But we can keep the issue open for now to collect further user reports, if any, or suggestions of heuristics we could try.

**Comment by Duckling92** [comment]
Thanks for the reply. Small correction on my side: `mypy` only flags the `SQLModel` case **when the Pydantic mypy plugin is enabled**. Without the plugin, `mypy` also lets the `SQLModel()` call pass.

With the plugin enabled, `mypy` *does* catch both cases, while `ty` only catches the plain `BaseModel` case - so the behavior still differs.

Minimal repro:

```python
from __future__ import annotations

from pydantic import BaseModel
from sqlmodel import SQLModel


class ExamplePydantic(BaseModel):
    number: float


class ExampleSQLModel(SQLModel):
    number: float


ExamplePydantic()   # E: Missing positional argument "number"
ExampleSQLModel()   # E: Missing positional argument "number"
```

`pyproject.toml`:

```ini
[tool.mypy]
plugins = ["pydantic.mypy"]
```

Command and output:

```bash
mypy example.py
example.py:15: error: Missing named argument "number" for "ExampleSQLModel"  [call-arg]
example.py:16: error: Missing named argument "number" for "ExamplePydantic"  [call-arg]
Found 2 errors in 1 file (checked 1 source file)
```

Environment:

* python 3.11.9
* pydantic 2.12.3
* sqlmodel 0.0.27
* mypy 1.18.2
* pydantic mypy plugin enabled

**Comment by carljm** [comment]
Thanks for the clarification. It's not clear to me exactly what the pydantic mypy plugin is doing that helps in this case, but it's something we can look at if we ever decide to build in similar special-cased support for pydantic.

**Comment by rayzeller** [comment]
> [🤖 ai-disclosure] 
> This response was drafted with Claude based on analysis of ty codebase and pyright's implementation.

---

hey, i took a crack at this in https://github.com/astral-sh/ruff/pull/22970 and wanted to share what i found.

after tracing through the code, i don't think the issue is about explicit `__init__` taking precedence. i think ty just wasn't *detecting* the `@dataclass_transform` at all for the SQLModel case.

here's the inheritance chain:
- `User` → `SQLModel` (metaclass=SQLModelMetaclass)
- `SQLModel` → `BaseModel` (metaclass=ModelMetaclass)
- `ModelMetaclass` is decorated with `@dataclass_transform`

when `code_generator_of_static_class()` iterates the MRO looking for dataclass-like bases, it was only calling `dataclass_transformer_params()` on each base - which checks if the class *itself* is decorated. it never called `try_metaclass()` to check if the base's metaclass has `@dataclass_transform`.

the fix adds that `try_metaclass()` check during MRO iteration. when we reach `BaseModel`, it finds `ModelMetaclass` which has the decorator.

per [PEP 681's metaclass example](https://peps.python.org/pep-0681/#metaclass-example), subclasses of a class using a `@dataclass_transform` metaclass should get dataclass-like semantics.

@erictraut - i noticed pyright does something similar at [dataClasses.ts lines 1460-1464](https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/analyzer/dataClasses.ts#L1460-L1464), checking `baseClass.shared.declaredMetaclass.shared.classDataClassTransform` during base class iteration. does this approach align with how pyright handles it?

**Comment by rayzeller** [comment]
separately, sqlmodel could also add `@dataclass_transform` directly to `SQLModelMetaclass`. that would make this work across all type checkers without requiring special detection logic. might be worth opening an issue on the sqlmodel repo suggesting this.

**Comment by sharkdp** [comment]
Let's not have this discussion in two places. Let's clarify this in https://github.com/astral-sh/ruff/pull/22970#issuecomment-3843164256 first.
